app/cluster/state/etcd.go

```
@@ -0,0 +1,279 @@
+	Workspaces string `json:"workspaces"` // comma separated workspaces
+	AckKey     string `json:"ack_key"`
+}
+
+type workspacesAckValue struct {
+	Status string `json:"status"`
+}
+
+func EnvETCDConfig() *ETCDConfig {
+	endpoints := strings.Split(config.GetEnv("ETCD_HOSTS", "127.0.0.1:2379"), `,`)
+	releaseName := config.GetEnv("RELEASE_NAME", `multitenantv1`)
+	serverIndex := config.GetInstanceID()
+
+	envConfigOnce.Do(func() {
+		config.RegisterDurationConfigVariable(time.Duration(15), &etcdGetTimeout, true, time.Second, "etcd.getTimeout")
+		config.RegisterDurationConfigVariable(time.Duration(3), &etcdWatchTimeout, true, time.Second, "etcd.watchTimeout")
+		config.RegisterDurationConfigVariable(time.Duration(30), &keepaliveTime, true, time.Second, "etcd.keepaliveTime")
+		config.RegisterDurationConfigVariable(time.Duration(10), &keepaliveTimeout, true, time.Second, "etcd.keepaliveTimeout")
+		config.RegisterDurationConfigVariable(time.Duration(20), &dialTimeout, true, time.Second, "etcd.dialTimeout")
+	})
+
+	return &ETCDConfig{
+		Endpoints:            endpoints,
+		ReleaseName:          releaseName,
+		ServerIndex:          serverIndex,
+		etcdWatchTimeout:     etcdWatchTimeout,
+		dialTimeout:          dialTimeout,
+		dialKeepAliveTime:    keepaliveTime,
+		dialKeepAliveTimeout: keepaliveTimeout,
+		etcdGetTimeout:       etcdGetTimeout,
+	}
+}
+
+type ETCDManager struct {
+	Config  *ETCDConfig
+	Client  *clientv3.Client
+	once    sync.Once
+	initErr error
+}
+
+func (manager *ETCDManager) init() error {
+	manager.once.Do(func() {
+		cli, err := clientv3.New(clientv3.Config{
+			Endpoints:            manager.Config.Endpoints,
+			DialTimeout:          manager.Config.dialTimeout,
+			DialKeepAliveTime:    manager.Config.dialKeepAliveTime,
+			DialKeepAliveTimeout: manager.Config.dialKeepAliveTimeout,
+			DialOptions: []grpc.DialOption{
+				grpc.WithBlock(), // block until the underlying connection is up
+			},
+		})
+		if err != nil {
+			manager.initErr = err
```

Why are we propogating an error which we encountered while trying to connect to etcd? Shouldn't we just exit the process and panic?

---

I would reserve panic for something much worse / unexpected. A communication err is bad, and we are going to exit gracefully, by propagating the error all the way to the top.